numba/targets/arrayobj.py

```diff
@@ -1839,6 +1839,81 @@ def np_unique_impl(a):
     return np_unique_impl
 
 
+@overload(np.repeat)
+def np_repeat(a, repeats):
+    # Implementation for repeats being a scalar is a module global function
+    # (see below) because it might be called from the implementation below.
+
+    def np_repeat_impl_repeats_array_like(a, repeats):
+        # implementation if repeats is an array like
+        repeats_array = np.asarray(repeats, dtype=np.int64)
+        # if it is a singleton array, invoke the scalar implementation
+        if repeats_array.shape[0] == 1:
+            return np_repeat_impl_repeats_scaler(a, repeats_array[0])
+        if np.any(repeats_array < 0):
+            raise ValueError("negative dimensions are not allowed")
+        asa = np.asarray(a)
+        aravel = asa.ravel()
+        n = aravel.shape[0]
+        if aravel.shape != repeats_array.shape:
+            raise ValueError(
+                "operands could not be broadcast together")
+        to_return = np.empty(np.sum(repeats_array), dtype=asa.dtype)
+        pos = 0
+        for i in range(n):
+            to_return[pos : pos + repeats_array[i]] = aravel[i]
+            pos += repeats_array[i]
+        return to_return
+
+    # type checking
+    if isinstance(a, (types.Array,
+                      types.List,
+                      types.BaseTuple,
+                      types.Number,
+                      types.Boolean,
+                      )
+                  ):
+        if isinstance(repeats, types.Integer):
+            return np_repeat_impl_repeats_scaler
+        elif isinstance(repeats, types.Float):
```

Comment:
Perhaps reorder these branches and alter the error messages as follows:

```python
if isinstance(repeats, types.Integer):
    return np_repeat_impl_repeats_scaler
elif isinstance(repeats, (types.Array, types.List)):
    if isinstance(repeats.dtype, types.Integer):
        return np_repeat_impl_repeats_array_like
    else:
        raise errors.TypingError("The repeats argument must be an integer")
else:
    raise errors.TypingError("The repeats argument must be an integer or an array-like of integer dtype")
```
so as to catch *anything* which is non-integer type and hint to the user what to do?

What do you think?

Reply:
sounds good, make it more readable too.

Reply:
I'm not sure the change made here: 65e2fc5 makes sense as it makes that error handling a catch-all e.g.
```python
from numba import njit
import numpy as np

def foo(a, r):
    return np.repeat(a, r)

x = ''
y = 1

print(njit(foo)(x, y))
```
now produces:
```
 TypingError: The repeats argument must be an integer or an array-like of integer dtype
```
whereas the problem is actually that `unicode_type` is not permitted for `a`.